### PR TITLE
Adds a TCP-first DNS lookup for joins.

### LINKED ISSUE #74: Perform on-join DNS requests over TCP
If we make the request via TCP we can more easily get a larger initial join list from the DNS server. Right now the default UDP client will only get a short list of responses (~3). If we pull SRV records we can get the port information as well.

### PR DESCRIPTION
**Pull request by slackpad**
This fixes #74.

### CODE REVIEW

File: memberlist.go [hunk 1]
```diff
@@ -191,50 +191,111 @@ func (m *Memberlist) Join(existing []string) (int, error) {
...
+	if len(cc.Servers) > 0 {
+		server := cc.Servers[0]
+		if !hasPort(server) {
+			server = net.JoinHostPort(server, cc.Port)
+		}
+
+		c := new(dns.Client)
+		c.Net = "tcp"
+		m := new(dns.Msg)
+		m.SetQuestion(host, dns.TypeANY)
```
Comment (sean-) [inline memberlist.go]: This comment is purely informational as there's nothing better to be done, but it's sad that this was brought up as a problem in '98 and it is still a problem (despite repeated attempts to fix over the years).  Item 4 in the following has additional detail.

http://tools.ietf.org/html/draft-sajitha-dnsext-qtype-addr-00

TL;DR: `ANY` is as good as it gets and works for the `A` and `AAAA` QTYPES, but it won't give the DNS server the opportunity to resolve `CNAME` records and return the correct information in the `ADDITIONAL` section.  Hold this thought for a second.

`T_ANY` also has the server xmit all records to the client, which shouldn't be common, and as you said, this is reasonably rare.

File: memberlist.go [hunk 2]
```diff
@@ -191,50 +191,111 @@ func (m *Memberlist) Join(existing []string) (int, error) {
...
+		}
+
+		var ips []ipPort
+		for _, r := range in.Answer {
+			switch rr := r.(type) {
+			case (*dns.A):
+				ips = append(ips, ipPort{rr.A, defaultPort})
+			case (*dns.AAAA):
+				ips = append(ips, ipPort{rr.AAAA, defaultPort})
+			}
```
Comment (sean-) [inline memberlist.go]: It might be worth handling `CNAME` or emitting a log message if there's a logger in scope about CNAME not being supported that response type comes back.  Intra-organization DNS shouldn't be using CNAMEs, but it's not unheard of.  The common case when talking to Consul doesn't have this problem, however.

File: memberlist.go [hunk 3]
```diff
@@ -191,50 +191,111 @@ func (m *Memberlist) Join(existing []string) (int, error) {
...
-
 	if numSuccess > 0 {
 		errs = nil
 	}
 	return numSuccess, errs
 }
 
+// ipPort holds information about a node we want to try to join.
+type ipPort struct {
+	ip   []byte
```
Comment (sean-) [inline memberlist.go]: Is there any reason the type shouldn't be `net.IP`?  `[]byte` is the backing store for `net.IP` so it should be equivalent.

**Comment by sean-** [comment]
LGTM

File: memberlist.go (hunk 3 above)
Comment (slackpad) [inline memberlist.go]: Good catch - I'll do that.

**Comment by slackpad** [comment]
@sean- thanks for the review. I didn't realize you were on it and force-pushed a change to add the “.” at the end if not there, and to not try local names without a “.” with the TCP resolver

File: memberlist.go (hunk 2 above)
Comment (slackpad) [inline memberlist.go]: I'll add a debug log about that (i could just emit a log for any other type of answer that i'm ignoring)

Comment (sean-) [inline memberlist.go]: That would be way too spammy (think of zones secured with `DNSSEC`).  IMO, the only `QTYPE` that's worth emitting a warning on is `CNAME` and everything else can be safely ignored.

**Comment by sean-** [comment]
Heh, no worries. GitHub is rather good about preserving comments across page reloads (props to whoever spent effort on making that happen at @github).

File: memberlist.go (hunk 2 above)
Comment (slackpad) [inline memberlist.go]: copy - will reduce to just `CNAME`
